### ISSUE TITLE
Request for Code to Compute Quantitative Results

### ISSUE THREAD

**Issue by miyugoko** [homangab/Track-2-Act#6]
Hello,
Thank you for releasing the code for inference and training. While I found implementations for qualitative results, I couldn’t locate the code for computing the quantitative evaluations reported in the paper. Could you release this part of the code?
Looking forward to your response.

**Comment by Magi329** [comment]
Hi, here is my code for quantitative evaluations. PS. the metric defined in his paper contains a "threshold" that renders the number invariant at the begining of training, so I change the metric to be the L1 distance of the pixels of the keypoints trajectories. Hoping to get some feedback on that should you try this metric ^^

```
## GLOBAL variable
ACTION_DIM = 800 ## 2* number of points
ACTION_HORIZON = 8 ## pred horizon


def compute_area_under_curve(pred_track, ground_truth, max_threshold, device):
    """
    Computes evaluation metric Δ = (sum_{t=1}^H sum_{x=1}^N δ^x_t) / H
    """
    delta = 0
    for threshold in range(1, max_threshold+1):
        delta += compute_avg_pixel_distance(pred_track, ground_truth, device)
    return delta / max_threshold   # the metric in his paper forgets to divide by max_threshold ??

def compute_avg_pixel_distance(pred_track, ground_truth, device):
    """
    Computes the average fraction δ of points that are within a threshold pixel
    distance of x of their ground truth in a time-step t.
    
    Parameters:
    - pred_track: A 2D array of shape (8, 400, 2), 
                        representing the predicted 8 steps of 400 points with their (x, y) pixel coordinates
    - ground_truth: A 2D array of shape (8, 400, 2), 
                        representing the ground truth 8 steps of 400 points with their (x, y) pixel coordinates
    - threshold (float): The pixel distance threshold for counting points as close to the ground truth.
    Returns:
    - float: The computed average fraction δ
    """   
    assert pred_track.shape == ground_truth.shape, "pred_track and ground_truth must have the same shape"

    timestep = len(pred_track)
    points_num = len(pred_track[0])

    distance = 0
    for t in range(timestep):
        distance_i = 0
        for i in range(points_num):
            distance_i += pixel_distance(pred_track[t][i], ground_truth[t][i], device)
        distance += distance_i/points_num

    return distance/timestep


def compute_fraction_within_threshold(pred_track, ground_truth, threshold, device):
    """
    Computes the average fraction δ of points that are within a threshold pixel
    distance of x of their ground truth in a time-step t.
    
    Parameters:
    - pred_track: A 2D array of shape (8, 400, 2), 
                        representing the predicted 8 steps of 400 points with their (x, y) pixel coordinates
    - ground_truth: A 2D array of shape (8, 400, 2), 
                        representing the ground truth 8 steps of 400 points with their (x, y) pixel coordinates
    - threshold (float): The pixel distance threshold for counting points as close to the ground truth.
    Returns:
    - float: The computed average fraction δ
    """   
    assert pred_track.shape == ground_truth.shape, "pred_track and ground_truth must have the same shape"

    timestep = len(pred_track)
    points_num = len(pred_track[0])

    within_t = 0
    for t in range(timestep):

        within_point = 0
        for i in range(points_num):
            if pixel_distance(pred_track[t][i], ground_truth[t][i], device) <= threshold:
                within_point += 1
        within_t += within_point/points_num

    return within_t/timestep


def pixel_distance(coord1, coord2, device):
    """
    Computes the L1 (Manhattan) distance between two pixel coordinates.
    """
    return torch.sum(torch.abs(coord1.to(device) - coord2.to(device)))


# normalize data
def get_data_stats(data):
    data = data.reshape(-1,data.shape[-1])
    stats = {
        'min': np.min(data, axis=0),
        'max': np.max(data, axis=0)
    }
    for i in range(len(stats['min'])):
        stats['min'][i] = 0
        stats['max'][i] = 96

    return stats

def normalize_data(data, stats):
    # nomalize to [0,1]
    for i in range(len(stats['min'])):
        stats['min'][i] = 0
        stats['max'][i] = 96
    ndata = (data - stats['min']) / (stats['max'] - stats['min'])
    # normalize to [-1, 1]
    ndata = ndata * 2 - 1
    return ndata


def evaluate_metric(model, data_loader, device):
    model.eval()  
    total_metric = 0
    total_samples = 0
    
    with torch.no_grad():  
        for nbatch in data_loader: 
            # get inputs and ground truth
            initial_image = nbatch['image'][0,:,0].squeeze().unsqueeze(0)   # [1, 3, 480, 640]
            goal_image = nbatch['image'][0,:,1].squeeze().unsqueeze(0)      # [1, 3, 480, 640]
            ground_truth = nbatch['action'][0].squeeze().unsqueeze(0)       # [1, 8, 800]

            # initialize action from Guassian noise
            z = torch.randn((1,ACTION_HORIZON,ACTION_DIM), device=device)
            x = z.clone()
            k = x[:,0,:] # prepare for image embed
            stats = {}
            stats['action'] = get_data_stats(z.detach().to('cpu').numpy())
            x = torch.from_numpy(normalize_data(x.detach().to('cpu').numpy(), 
            stats=stats['action'])).float().to(device)
            x = torch.reshape(x,(x.shape[0],ACTION_HORIZON,int(ACTION_DIM/2),2))
            x = torch.swapaxes(x,1,2)
            x = torch.reshape(x,(x.shape[0],int(ACTION_DIM/2),16))                
            x = torch.swapaxes(x,1,2)
            z = torch.randn((x.shape[0],x.shape[1] , x.shape[2]), device=device)

            # images
            y = torch.cat((initial_image, goal_image), dim=0).to(device)
            y = y.float()  # .permute(0, 3, 1, 2)
            y=y[None,:].to(device)
            model_kwargs = dict(y=y)

            # k
            k = torch.from_numpy(normalize_data(k.detach().to('cpu').numpy(), 
            stats=stats['action'])).float().to(device)
            k = k[None,:]
            k = torch.reshape(k,(k.shape[0],1,int(ACTION_DIM/2),2))
            k = torch.swapaxes(k,1,2)
            k = torch.reshape(k,(k.shape[0],int(ACTION_DIM/2),2))                
            k = torch.swapaxes(k,1,2)


            diffusion = create_diffusion(str(250)) # default

            naction = diffusion.p_sample_loop(model.forward, z.shape, z, clip_denoised=False, 
            model_kwargs=model_kwargs, progress=True, device=device,point_conditioned=True,img=k)

            # change shape
            naction = torch.swapaxes(naction,1,2)
            naction = torch.reshape(naction,(naction.shape[0],int(ACTION_DIM/2),ACTION_HORIZON,2))
            naction = torch.swapaxes(naction,1,2)
            pred_track = torch.reshape(naction,(naction.shape[0],ACTION_HORIZON,int(ACTION_DIM)))

            
            # compute metric
            metric = compute_area_under_curve(pred_track, ground_truth, 10, device) 
            total_metric += metric    
            total_samples += 1        

 
    avg_metric = total_metric / total_samples
    model.train()  
    return avg_metric
```
